tests/python/pants_test/pantsd/test_pantsd_integration.py

```diff
@@ -462,16 +464,25 @@ def _assert_pantsd_keyboardinterrupt_signal(self, signum):
         config,
       )
 
-      time.sleep(5)
-      checker.assert_started()
+      # TODO: There is a race condition here, where if the waiter run is killed before it starts the
+      # remote client, it will display the PantsRunnerSignalHandler message, but if it is signalled
+      # after switching control to the remote client, it will display a message about being
+      # interrupted via the pailgun client (which is what we check here). This means we aren't
+      # necessarily testing that the remote client is handling signals correctly, but it does mean
+      # the test won't flake until we can figure out a better method than relying on timeouts. This
+      # 16-second sleep is longer than the timeout we wait for the process to start, and is
+      # *hopefully* long enough to have pants transition to the remote client phase.
+      time.sleep(16)
```

Comment:
Woah. Can you at least file a bug about fixing this?

But also, let's talk about it here... I don't think I entirely follow the bug that means this is necessary. What error message would we actually see if we removed this code?

Reply:
The alternative is to check for one of two error messages -- the one we check here, as well as the one that is set in `pants_runner.py`'s signal handler (which raises if we don't get to invoking the pailgun yet in the pants subprocess before the signal is sent). We already check the `pants_runner.py` signal handling message in our exception sink integration testing, so this was an attempt to goad pants to fail in the way we expect. Timeouts shouldn't be used here, this is the cause of prior flakiness.

Reply:
I think increasing the timeout for awaiting the pid in `assert_started()` *might* make sense still, although it's probably worth looking into why it takes more than a single second to write the pid file. In this case it might also make sense to have some explicit IPC from the pants subprocess back to the parent pants process (which might also just take the form of writing a file) to indicate that control has shifted to the remote client connecting to pantsd.

Reply:
Also, if we kept pantsd alive between integration tests, we could remove all of these timeouts, since none of this testing relates to starting up pantsd itself, just about the remote client connecting to it.

Reply:
It feels like removing this `time.sleep` and just letting the `checker.assert_started` do the waiting in a polling way would be sufficient here, right?

Reply:
Correct, @blorente helped me realize this yesterday.